### PR TITLE
hikey960: enable IOMCU DMAC

### PR DESCRIPTION
**Pull request by oscardagrach**
There exists a third DMA controller on the hi3660
SoC called the IOMCU DMAC. This controller is used by
peripherals like SPI2 and UART3. Initialize channels 4-7
as non-secure, while 0-3 remain reserved and secure.

Signed-off-by: Ryan Grachek <ryan@edited.us>

### CODE REVIEW

File: plat/hisilicon/hikey960/include/hi3660.h [hunk 1]
```diff
@@ -373,4 +373,11 @@
 #define EDMAC_SEC_CTRL_GLOBAL_SEC		(1 << 0)
 #define EDMAC_CHANNEL_NUMS				16
 
+#define IOMCU_DMAC_BASE			0xffd77000
+#define IOMCU_DMAC_SEC_CTRL		(IOMCU_DMAC_BASE + 0x694)
+#define IOMCU_DMAC_AXI_CONF(x)			(IOMCU_DMAC_BASE + 0x820 + (x << 6))
```
Comment (antonio-nino-diaz-arm) [inline plat/hisilicon/hikey960/include/hi3660.h]: ``x`` should be guarded by ``()``

File: plat/hisilicon/hikey960/hikey960_bl31_setup.c [hunk 2]
```diff
@@ -141,6 +141,20 @@ static void hikey960_edma_init(void)
...
+{
+	int i;
+	uint32_t non_secure;
+
+	non_secure = IOMCU_DMAC_SEC_CTRL_INTR_SEC | IOMCU_DMAC_SEC_CTRL_GLOBAL_SEC;
+	mmio_write_32(IOMCU_DMAC_SEC_CTRL, non_secure);
+
+	/* channels 0-3 are reserved */
+	for (i = 4; i < IOMCU_DMAC_CHANNEL_NUMS; i++) {
+		mmio_write_32(IOMCU_DMAC_AXI_CONF(i), (1 << 6) | (1 << 18));
```
Comment (antonio-nino-diaz-arm) [inline plat/hisilicon/hikey960/hikey960_bl31_setup.c]: Don't use magic numbers if you can. Add the definitions to the header file along with the others.

**Comment by antonio-nino-diaz-arm** [comment]
IP Review +1

File: plat/hisilicon/hikey960/hikey960_bl31_setup.c (hunk 2 above)
Comment (oscardagrach) [reply plat/hisilicon/hikey960/hikey960_bl31_setup.c]: These register are not publicly documented by Hisilicon, but I know them to be standard AXI bus access permissions. I will add the definitions.

**Comment by johnstultz-work** [comment]
@xu-yiping Can you take a look at this as well and provide any review feedback?

**Comment by hzhuang1** [comment]
Acked.

**Comment by antonio-nino-diaz-arm** [comment]
jenkins: test this please

**Comment by antonio-nino-diaz-arm** [comment]
The failures are MISRA defects, the patch is ok.

**Comment by xu-yiping** [comment]
looks good to me

**Comment by xu-yiping** [comment]
looks good to me

**Comment by jforissier** [comment]
Hi,

This patch seems to break the boot on my HiKey960 board. Boot log:
```
hikey960 boarid:5301 xloader use UART6
scsysstat_value[0].
clear reset source
last_keypoint0,reboot_type0
secdbg not DCU.
SecDbgVer exit

 xloader chipid is: 0x36600110, start at 438ms.
Build Date: Dec  6 2017, 15:31:59
[clock_init] ++
hikey960 [hikey960_clk_init]
hi3660 [clk_setup]
[clock_init] --
storage type is UFS
ufs retry: 6 count v_tx:0 v_rx:0
ufs set v_tx:0 v_rx:0
Hikey960[5301] no need avs_init.
ddr ft:0xf20332a3,mode:1 target:4
UceLdOk
ch 0 gt_errfail, STATUS:0x00000060
ch 0 gdst_errfail, STATUS:0x00000040
ch 1 gt_errfail, STATUS:0x00000060
ch 1 gdst_errfail, STATUS:0x00000040
ch 2 gt_errfail, STATUS:0x00000060
ch 2 gdst_errfail, STATUS:0x00000040
ch 3 gt_errfail, STATUS:0x00000060
ch 3 gdst_errfail, STATUS:0x00000040
timeout
timeout
timeout
timeout
density: 0x0c0c0c0c,0x00000000,0x0c0c0c0c,0x00000000,0x0c0c0c0c,0x00000000,0x0c0c0c0c,0x00000000
ddr info 0x00000306
400M
685M
1067M
C0R,V0x0000002c e:113
C1R,V0x0000002d e:66
C2R,V0x0000002c e:113
C0R,V0x0000002d e:66
C1R,V0x0000002e e:66
C2R,V0x0000002d e:66
C3R,V0x0000002d e:66
C0R,V0x0000002e e:66
C1R,V0x0000002f e:66
C2R,V0x0000002e e:66
C3R,V0x0000002e e:66
C0R,V0x0000002f e:65
C1R,V0x00000030 e:65
C2R,V0x0000002f e:65
C3R,V0x0000002f e:66
1244M
1866M
C3R,V0x00000016 e:113
C0R,V0x00000017 e:66
C2R,V0x00000017 e:66
C3R,V0x00000017 e:66
iomcu_subsys_init
boot_c0 PROFILE 4
slave0 irq0:0x00000004
slave1 irq0:0x00000004
NOTICE:  BL2: v2.1(release):v2.2-rc0
NOTICE:  BL2: Built : 10:42:45, Oct 15 2019
NOTICE:  BL2: Booting BL31
NOTICE:  BL31: v2.1(release):v2.2-rc0-dirty
NOTICE:  BL31: Built : 10:55:25, Oct 15 2019
NOTICE:  plat/hisilicon/hikey960/hikey960_bl31_setup.c:149
NOTICE:  plat/hisilicon/hikey960/hikey960_bl31_setup.c:151
```

Patch:
```diff
trusted-firmware-a ((v2.2-rc0))$ git diff
diff --git a/plat/hisilicon/hikey960/hikey960_bl31_setup.c b/plat/hisilicon/hikey960/hikey960_bl31_setup.c
index 9383265e..38e74694 100644
--- a/plat/hisilicon/hikey960/hikey960_bl31_setup.c
+++ b/plat/hisilicon/hikey960/hikey960_bl31_setup.c
@@ -146,13 +146,18 @@ static void hikey960_iomcu_dma_init(void)
        int i;
        uint32_t non_secure;

+       NOTICE("%s:%d\n", __FILE__, __LINE__);
        non_secure = IOMCU_DMAC_SEC_CTRL_INTR_SEC | IOMCU_DMAC_SEC_CTRL_GLOBAL_SEC;
+       NOTICE("%s:%d\n", __FILE__, __LINE__);  /* This is line 151 */
        mmio_write_32(IOMCU_DMAC_SEC_CTRL, non_secure);
+       NOTICE("%s:%d\n", __FILE__, __LINE__);

        /* channels 0-3 are reserved */
        for (i = 4; i < IOMCU_DMAC_CHANNEL_NUMS; i++) {
+               NOTICE("%s:%di i=%d\n", __FILE__, __LINE__, i);
                mmio_write_32(IOMCU_DMAC_AXI_CONF(i), IOMCU_DMAC_AXI_CONF_ARPROT_NS |
                                 IOMCU_DMAC_AXI_CONF_AWPROT_NS);
+               NOTICE("%s:%di i=%d\n", __FILE__, __LINE__, i);
        }
 }
```

If I revert commit c8ab47d42db4 ("hikey960: enable IOMCU DMAC") on top of 2.2-rc0, the board boots again.

Any ideas how to investigate/fix this?

**Comment by oscardagrach** [comment]
When did you notice the regression?

Is there new SCP firmware or Xloader?

I can't tell very much without a debug log from ATF, but it looks like
something is configuring
the IOMCU DMAC AXI bus permissions prior to BL31/ATF, but BL31 should
be accessing the registers
through secure context so this shouldn't be an issue, however I am not sure.

This really shouldn't be something BL31 should be hanging up on, it's
just accessing two MMIO registers and writing a dword
to each. Something must have changed elsewhere.

Did you happen to be using the Hisilicon bootchain at any point prior
to using ATF?

**Comment by jforissier** [comment]
Hi @oscardagrach,

> When did you notice the regression?

A few of days ago. I was running some tests in preparation for OP-TEE release 3.7.0, and I thought I would take the opportunity to try and update TF-A. We are currently using v2.0-507-g34efb683 [1] which dates back from 9 months ago. The build environment is described at [1].

[1] https://optee.readthedocs.io/en/latest/building/devices/hikey960.html

> Is there new SCP firmware or Xloader?

No, not at all. What I did was:
1. Build and run everything once (`make toolchains`, `make -j10`, `make flash`)
2. Checkout TF-A tag v2.2-rc0
3. Rebuild everything (`make -j10`, `make flash`)
4. At this point the board won't boot past the BL31 message
5. Checkout the previous TF-A version (or just revert the IOMCU DMA commit)
6. Build again and use recovery procedure to flash the board (`make -j10`, `make recovery`)
7. Power off/on: the board boots and works fine

> Did you happen to be using the Hisilicon bootchain at any point prior
to using ATF?

I'm not sure about this, I guess the answer is "yes" because I sometimes need to use the recovery procedure (when OP-TEE init is stuck for example), in which case the following files are used: `hisi-sec_usb_xloader.img`, `hisi-sec_uce_boot.img`, `recovery.bin`.

**Comment by hzhuang1** [comment]
Hi @jforissier 

I've updated OPTEE to 3.7.0-rc1 and ATF to latest code. But I can't see the same issue that you mentioned. It boots successfully on my hikey960 platform.

**Comment by jforissier** [comment]
@hzhuang1 thanks for testing! FYI the procedure I have followed:
```
mkdir optee_repo_hikey960
cd optee_repo_hikey960
repo init -u https://github.com/OP-TEE/manifest.git -m hikey960.xml
repo sync -j20 --no-clone-bundle
cd trusted-firmware-a
git checkout v2.2-rc0
cd ../build
make toolchains
# Switch board to recovery mode (ON-ON-OFF), power it on
make recovery
# Power off board, switch it to normal boot mode (ON-OFF-OFF), power it on
# => hangs at: "NOTICE:  BL31: Built : 08:42:13, Oct 16 2019"
```
(to be honest when I ran the test again, I built on a remote host due to network performance issues on my local laptop, then I transferred the binaries locally for flashing -- hopefully I did not mess things up...)

**Comment by hzhuang1** [comment]
@jforissier I can reproduced it with your instructions. But I can't reproduce it with my build scripts. Since code base are same, I doubt it's related to toolchain. I find you're using gcc 8.3. How to switch toolchain in your repo?

**Comment by hzhuang1** [comment]
> @jforissier I can reproduced it with your instructions. But I can't reproduce it with my build scripts. Since code base are same, I doubt it's related to toolchain. I find you're using gcc 8.3. How to switch toolchain in your repo?

If this issue could be avoided with other toolchain in your repo, it could be simplified as instruction sequences out of order. Then we need some memory barrier in this patch.

**Comment by jforissier** [comment]
@hzhuang1 thank you so much for taking time to investigate this :+1:

Yes I'm using 8.3, can be changed by symlinking `toolchains/aarch64` to another cross-compiler (the command used during build is `toolchains/aarch64/bin/arm-linux-gnueabihf-gcc` etc.).

I have GCC 6.2.1 on my machine, let me try it.

**Comment by jforissier** [comment]
@hzhuang1 I tried GCC 6.2.1 (with `br-ext/configs/toolchain-aarch` reverted to version 53119b7^ to avoid a build error) => same issue, does not boot :-/

**Comment by hzhuang1** [comment]
@jforissier I downloaded gcc 8.3 and integrated into my build scripts. It could boot successfully with DEBUG build. But I even met an error in ATF in recovery mode with RELEASE build.
As a workaround, I suggest you use gcc 7.2.1 or some version instead.
Maybe appending memory barrier could fix this issue, it also costs a lot of time to debug.

**Comment by hzhuang1** [comment]
> @jforissier I downloaded gcc 8.3 and integrated into my build scripts. It could boot successfully with DEBUG build. But I even met an error in ATF in recovery mode with RELEASE build.
> As a workaround, I suggest you use gcc 7.2.1 or some version instead.
> Maybe appending memory barrier could fix this issue, it also costs a lot of time to debug.

I don't recommend to use gcc 5.x or gcc 6.x. Since there're some issues. The well verified toolchain is 4.8/4.9 and 7.2.1/7.x.

**Comment by jforissier** [comment]
>  It could boot successfully with DEBUG build. 

Weird. Debug build does not work for me (whether GCC 6.2.1, 7.2.1 or 8.3.0).

> Maybe appending memory barrier could fix this issue, it also costs a lot of time to debug.

By chance, do you have access to a JTAG debugger? I have used OpenOCD in the past, but it is not super reliable, and in this case I could not do anything useful :-/

**Pull request merged**